[Documentation] Add AppGateServer Docs (#200)
pokt-network/poktroll: pkg/appgateserver/docs/server.md

# AppGateServer

> <!-- toc -->

Author: markdown-toc for neovim

```

When started without the `self-signing` flag the `AppGateServer` will not store
the address associated with the `signing-key` flag. This means the requests that
it receives **must** include the application's address in the
[endpoint](#requests).

When a gateway receives a request it will fetch the ring of the application
sending the request. It will then sign the request using `signing-key` used to
start the gateway.

_NOTE: The gateway is only able to sign relays for applications who have
delegated to the gateway_

## Requests

Requests from an application are received on the `AppGateServer`'s listening
address. This acts as a single endpoint the application can send their requests
to.

When sending a request the service **must** be specified in the endpoint and
if the `AppGateServer` is a gateway then the application's address must also be
included.

The endpoint will look like the following:

```
scheme://host:port/{service}[?senderAddr={application address}]
```

### Request Types

Depending on the service the request will be of a different form: JSON-RPC,
REST, gRPC, GraphQL, etc. These different payload types can be catagorised into
two catagories.

> 1. Symmetric Protocols

Reviewer: @red-0ne @h5law **IMPORTANT COMMENT.**

I think this terminology is important so I wanted to understand if this is something we coined or if it's a standard?

---

[1] https://en.wikipedia.org/wiki/RPyC

I could insinuate was symmetric/assymetric protocols are but did a bit of research to see if it conflicts with industry standards. [1]
![Screenshot 2023-11-20 at 8 01 20 PM](https://github.com/pokt-network/poktroll/assets/1892194/30904d62-d636-46a7-926f-8da40cfd595a)
![Screenshot 2023-11-20 at 7 59 20 PM](https://github.com/pokt-network/poktroll/assets/1892194/7ab522b9-590b-4a00-99a2-7cb978172501)

I think we should call it `Synchronous` and `Asynchronous`

![Screenshot 2023-11-20 at 8 03 48 PM](https://github.com/pokt-network/poktroll/assets/1892194/522fb6aa-a30f-43d4-9772-c17e2be41d94)

is used for other types of symmetric protocols, extracting different fields for
each payload type.

#### Asymmetric Protocols

_NOTE: Asymmetric Protocols are currently unsupported as of writing this
document_

## Sessions

The `AppGateServer` will handle the interaction with the chain to retrieve the
current session for the application whose request it is handling. This enables
the application to simply send their requests without having to manage their
sessions or keep track of them.

The `AppGateServer` will query the chain for the current session ID for the
given application and service pair.

See [session.go](../session.go) for more details on the implementation of
session retrieval.

## Endpoint Selection

Endpoints for the suppliers in a session are currently chosen based on the
first suitable supplier for the service from within the session.

_NOTE: The endpoint selection algorithm is planned to factor in other factors
such as Quality of Service and move away from the greedy selection
implementation_

See [endpoint-selector.go](../endpoint_selector.go) for more details on the
implementation of supplier endpoint selection.

## Signatures

> The `AppGateServer` signs all requests (as both applications and gateways) using

Reviewer 2: ```suggestion
The `AppGateServer` signs all requests (in both application and gateway mode) using
```

the ring of the application sending the request. They sign the ring using the
private key associated with the key used when starting the server.

When a response is received from the supplier, it is verified using the public

> key of the supplier who made the request.

Reviewer 2: ```suggestion
key of the supplier who executed the request.
```

### Rings

> Rings are created using the application's on-chain state. This includes a list

Reviewer 2: ```suggestion
Rings are created using the application's on-chain state. In gateway mode, this includes a list
```

Author: This will always include the gateways, it depends on whether the application has delegated or not, but if it has and it starts its own appgateserver to self sign it will still have those in its ring
